warehouse/api/integrity.py

```diff
@@ -0,0 +1,80 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pyramid.httpexceptions import HTTPForbidden, HTTPNotAcceptable, HTTPNotFound
+from pyramid.request import Request
+from pyramid.view import view_config
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.cache.http import add_vary, cache_control
+from warehouse.cache.origin import origin_cache
+from warehouse.packaging.models import File
+from warehouse.utils.cors import _CORS_HEADERS
+
+MIME_TEXT_HTML = "text/html"
+MIME_PYPI_INTEGRITY_V1_HTML = "application/vnd.pypi.integrity.v1+html"
+MIME_PYPI_INTEGRITY_V1_JSON = "application/vnd.pypi.integrity.v1+json"
+
+
+def _select_content_type(request: Request) -> str:
+    offers = request.accept.acceptable_offers(
+        [
+            # JSON currently has the highest priority.
+            MIME_PYPI_INTEGRITY_V1_JSON,
+            MIME_TEXT_HTML,
+            MIME_PYPI_INTEGRITY_V1_HTML,
+        ]
+    )
+
+    # Default case: JSON.
+    if not offers:
+        return MIME_PYPI_INTEGRITY_V1_JSON
+    else:
+        return offers[0][0]
+
+
+@view_config(
+    route_name="attestations.provenance",
+    context=File,
+    require_methods=["GET"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+    decorator=[
+        add_vary("Accept"),
+        cache_control(10 * 60),  # 10 minutes
+        origin_cache(
+            1 * 24 * 60 * 60,  # 1 day
+            stale_while_revalidate=5 * 60,  # 5 minutes
+            stale_if_error=1 * 24 * 60 * 60,  # 1 day
+        ),
+    ],
+)
+def provenance_for_file(file: File, request: Request):
+    # Determine our response content-type. For the time being, only the JSON
+    # type is accepted.
+    request.response.content_type = _select_content_type(request)
+    if request.response.content_type != MIME_PYPI_INTEGRITY_V1_JSON:
+        return HTTPNotAcceptable(json={"message": "Request not acceptable"})
+
+    if request.flags.enabled(AdminFlagValue.DISABLE_PEP740):
+        return HTTPForbidden(json={"message": "Attestations temporarily disabled"})
+
+    if not file.provenance:
+        return HTTPNotFound(
+            json={"message": f"No provenance available for {file.filename}"}
+        )
+
+    # Apply CORS headers.
+    request.response.headers.update(_CORS_HEADERS)
+
+    return file.provenance.provenance
```

Comment:
Seeing this made me double-take, since I thought it must be a typo. Is there a better name for `Provenance.provenance`, maybe `Provenance.attestations_json` since that's what they contain? 

Then this line would read:

```python
return file.provenance.attestations_json
```

Is the naming confusing, or is this the normal naming patterns of this universe?

Reply:
Yeah, this naming is non-ideal. The reason I named it `provenance` instead of `attestations_json` is because a "provenance object" is more than just a set of "attestation objects" in PEP 740 (it also has groupings/associated identity information).

WDYT about `provenance_json`? If that seems reasonable to you, I'll pin this for a follow-up PR (since it'll require a small migration).

Reply:
Well, a column rename from `provenance` to `provenance_json` doesn't seem to be worth the churn, so it's probably fine as is.

Although if and when you might need to do any migrations related to the Provenance model, please add `comment="..."` to the `mapped_columm` definition, so at the very least the comments appear in the database as well. 

See examples for a table itself as well as all the columns here:
https://github.com/pypi/warehouse/blob/7273c870bafe5d677d57dfdcbe70c86e5ab53a79/warehouse/ip_addresses/models.py#L31-L67